### ISSUE TITLE
Non deterministic failure of GFN-FF

### ISSUE THREAD

**Issue by awvwgk** [grimme-lab/xtb#276]
**Describe the bug**
Currently some test cases show non-deterministic behaviour with the GFN-FF method.
This seems to be limited to GCC compiled binaries, as the error cannot be observed with Intel compiled ones.

**To Reproduce**
Steps to reproduce the behaviour:

1. Checkout current master (dc4c10ea9191160a39cd5ee62e8695ab2a9e6edc).
2. Enable the `remdesivir` test in the GFN-FF scale-up test:

```patch
diff --git a/TESTSUITE/gfnff.f90 b/TESTSUITE/gfnff.f90
index 060f64e..c4f4242 100644
--- a/TESTSUITE/gfnff.f90
+++ b/TESTSUITE/gfnff.f90
@@ -464,7 +464,7 @@ subroutine test_gfnff_scaleup
       & 0.12497794198721_wp, 0.19334610335377_wp]
 
    call init(env)
-   do iMol = 1, 5, 2
+   do iMol = 1, 5
       if (afail > 0) exit
 
       call getMolecule(mol, trim(molecules(iMol)))
```

3. Enter the build directory and repeat the Scale-up test several times to observe a failure

```
meson test 'GFN-FF: Scale-up' --repeat 100
...
Ok:                 75  
Expected Fail:      0   
Fail:               25  
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0
```

The energy/gradient missmatch is shown once the `-v` flag is added (don't add the flag when running repeats or it will flood your terminal, just repeat the test until you get a failure):

```
assertion:  -13.822813528701     ==  -13.822813542638     FAILED
assertion:  0.12497792627240     ==  0.12497794198721     FAILED
```

The error originates from the Halogen bonding contribution:

```diff
--- fail.txt	2020-06-24 09:51:45.687824147 +0200
+++ good.txt	2020-06-24 09:52:15.619153851 +0200
@@ -150,7 +150,7 @@
          :: electrostat energy        -0.567666318775 Eh    ::
          :: dispersion energy         -0.084209154309 Eh    ::
          :: HB energy                 -0.004454104996 Eh    ::
-         :: XB energy                 -0.004644053515 Eh    ::
+         :: XB energy                 -0.004644067452 Eh    ::
          :: bonded atm energy         -0.011556750556 Eh    ::
          :: external energy            0.000000000000 Eh    ::
          :: -> Gsolv                  -0.077654004103 Eh    ::
```

Output files to verify:

[good.txt](https://github.com/grimme-lab/xtb/files/4823892/good.txt)
[fail.txt](https://github.com/grimme-lab/xtb/files/4823893/fail.txt)

**Expected behaviour**
Deterministic behaviour of the Halogen bonding contribution, better even correct behaviour of the contribution.

**Additional Details**

Tested on a Manjaro Linux with GCC and MKL2020 backend:

```
gfortran -v
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.1.0 (GCC)
```

@sespic any idea how to fix this?

**Comment by timostrunk** [comment]
I found a possible reason, but only a bandaid solution, which (in my opinion) might be just masking another bug.

The problem can be worked around by setting sqrab = 0 in src/gfnff/gfnff_eg.f90, line 135:

      ghb = 0.0d0
      gshift = 0.0d0

      allocate(sqrab(n*(n+1)/2),srab(n*(n+1)/2),qtmp(n),g5tmp(3,n), &
     &         eeqtmp(2,n*(n+1)/2),d3list(2,n*(n+1)/2),dcn(3,n,n),cn(n), &
     &         hb_dcn(3,n,n),hb_cn(n))
      sqrab = 0

Before doing this, I get:
Ok:                 70  
Expected Fail:      0   
Fail:               30  
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

After doing this, I get:
Ok:                 100 
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

It explains, why the fail is compiler dependent. One compiler will have 0 initialized the field, the other not.

If it is not initialized, the if in src/gfnff/gfnff_ini2.f90 line 753
        if(rab.gt.hbthr2)cycle 

trips up sometimes in Valgrind, because rab is unitialized. My best guess would be that an index ij=j+i*(i-1)/2 is odd maybe because of odd values in topo%xbatABl, but I did not investigate further.

**Comment by awvwgk** [comment]
Thanks. This seems to narrow the problem down, the `topo%xbatABl` is by no means constrained to the the lower triangle, which means obtaining the composite index with `ij=j+i*(i-1)/2` is not correct.

If this is the case, I still don't understand why it is not *failing deterministically* than.

**Comment by timostrunk** [comment]
This is just an observation from me [citation needed]:

Very often, when I allocate memory, even though it is not initialized, it is filled with zeroes, i.e. equivalent to my quickfix and most probably equivalent to the intel compiler, which I cannot test. It would only fail deterministically, if the memory block always contained garbage, which I can simulate with the line:
gfn_ffini2.f90:135

      sqrab = 1232323211

In this case, I get a 100% fail rate.

**Comment by awvwgk** [comment]
The error seems to related to the way meson is running unit tests, see https://mesonbuild.com/Reference-manual.html#test. Since the error actually never happens in production runs and is tied to the usage of `MALLOC_PERTURB_` it is not an urgent problem to fix.

Nevertheless, we will keep in eye on this, as it is a hint for an issue with the underlying implementation.

**Comment by timostrunk** [comment]
Please consider reopening this bug report. It is a very severe bug, because it is hard to detect. Turning off memory perturbation will make it even harder to detect in the future.

It is expected that it rarely triggers due to 0-initialization of memory, when no memory perturbation is present. But the Fortran standard does not force arrays to be 0-initialized.

After the bug was closed I looked into it again and found it to be the missing initialization of the diagonal of a distance matrix. It is fixed in #308

**Comment by awvwgk** [comment]
Agreed, a real fix is better than some hack for this issue.
